presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+ * <p>
+ * Based on: https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/type/UnsignedInt128.java
+ * and https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/type/Decimal128.java
+ */
+public final class UnscaledDecimal128Arithmetic
+{
+    public static final int UNSCALED_DECIMAL_128_SLICE_LENGTH = 2 * SIZE_OF_LONG;
+
+    private static final Slice[] POWERS_OF_TEN = new Slice[MAX_PRECISION];
+
+    private static final int SIGN_LONG_INDEX = 1;
+    private static final int SIGN_INT_INDEX = 3;
+    private static final long SIGN_LONG_MASK = 1L << 63;
+    private static final int SIGN_INT_MASK = 1 << 31;
+    private static final int SIGN_BYTE_MASK = 1 << 7;
+    private static final int NUMBER_OF_LONGS = 2;
+    private static final long ALL_BITS_SET_64 = 0xFFFFFFFFFFFFFFFFL;
+    /**
+     * Mask to convert signed integer to unsigned long.
+     */
+    private static final long LONG_MASK = 0xFFFFFFFFL;
+
+    /**
+     * 5^13 fits in 2^31.
+     */
+    private static final int MAX_POWER_OF_FIVE_INT = 13;
+    /**
+     * 5^x. All unsigned values.
+     */
+    private static final int[] POWERS_OF_FIVES_INT = new int[MAX_POWER_OF_FIVE_INT + 1];
+
+    /**
+     * 10^9 fits in 2^31.
+     */
+    private static final int MAX_POWER_OF_TEN_INT = 9;
+    /**
+     * 10^18 fits in 2^63.
+     */
+    private static final int MAX_POWER_OF_TEN_LONG = 18;
+    /**
+     * 10^x. All unsigned values.
+     */
+    private static final int[] POWERS_OF_TEN_INT = new int[MAX_POWER_OF_TEN_INT + 1];
+
+    private static final Unsafe unsafe;
+
+    static {
+        try {
+            // fetch theUnsafe object
+            Field field = Unsafe.class.getDeclaredField("theUnsafe");
+            field.setAccessible(true);
+            unsafe = (Unsafe) field.get(null);
+            if (unsafe == null) {
+                throw new RuntimeException("Unsafe access not available");
+            }
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        for (int i = 0; i < POWERS_OF_TEN.length; ++i) {
+            POWERS_OF_TEN[i] = unscaledDecimal(BigInteger.TEN.pow(i));
+        }
+
+        POWERS_OF_FIVES_INT[0] = 1;
+        for (int i = 1; i < POWERS_OF_FIVES_INT.length; ++i) {
+            POWERS_OF_FIVES_INT[i] = POWERS_OF_FIVES_INT[i - 1] * 5;
+        }
+
+        POWERS_OF_TEN_INT[0] = 1;
+        for (int i = 1; i < POWERS_OF_TEN_INT.length; ++i) {
+            POWERS_OF_TEN_INT[i] = POWERS_OF_TEN_INT[i - 1] * 10;
+        }
+
+        if (ByteOrder.nativeOrder().equals(ByteOrder.BIG_ENDIAN)) {
+            throw new IllegalStateException("UnsignedDecimal128Arithmetic is supported on little-endian machines only");
+        }
+    }
+
+    public static Slice unscaledDecimal()
+    {
+        return Slices.allocate(UNSCALED_DECIMAL_128_SLICE_LENGTH);
+    }
+
+    public static Slice unscaledDecimal(Slice decimal)
+    {
```

Comment:
Input decimals should have 4 ints. We skip length assertion to reduce CPU branching during computations.